numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1434 @@
+ * if it cannot and DType is used instead, a cast will have to be tried.
+ *
+ * @param fixed_DType A user provided fixed DType, can be NULL
+ * @param DType A discovered DType (by discover_dtype_from_pyobject);
+ *        this can be identical to `fixed_DType`, if it obj is a
+ *        known scalar. Can be `NULL` indicating no known type.
+ * @param obj The Python scalar object. At the time of calling this function
+ *        it must be known that `obj` should represent a scalar.
+ */
+static NPY_INLINE PyArray_Descr *
+find_scalar_descriptor(
+        PyArray_DTypeMeta *fixed_DType, PyArray_DTypeMeta *DType,
+        PyObject *obj)
+{
+    PyArray_Descr *descr;
+
+    if (DType == NULL && fixed_DType == NULL) {
+        /* No known DType and no fixed one means we go to object. */
+        return PyArray_DescrFromType(NPY_OBJECT);
+    }
+    else if (DType == NULL) {
+        /*
+         * If no DType is known/found, give the fixed give one a second
+         * chance.  This allows for example string, to call `str(obj)` to
+         * figure out the length for arbitrary objects.
+         */
+        descr = fixed_DType->discover_descr_from_pyobject(fixed_DType, obj);
+    }
+    else {
+        descr = DType->discover_descr_from_pyobject(DType, obj);
+    }
+    if (descr == NULL) {
+        return NULL;
+    }
+    if (fixed_DType == NULL) {
+        return descr;
+    }
+
+    Py_SETREF(descr, cast_descriptor_to_fixed_dtype(descr, fixed_DType));
+    return descr;
+}
+
+
+/**
+ * Assign a single element in an array from a python value.
+ *
+ * The dtypes SETITEM should only be trusted to generally do the right
+ * thing if something is known to be a scalar *and* is of a python type known
+ * to the DType (which should include all basic Python math types), but in
+ * general a cast may be necessary.
+ * This function handles the cast, which is for example hit when assigning
+ * a float128 to complex128.
+ *
+ * At this time, this function does not support arrays (historically we
+ * mainly supported arrays through `__float__()`, etc.). Such support should
+ * possibly be added (although when called from `PyArray_AssignFromCache`
+ * the input cannot be an array).
+ * Note that this is also problematic for some array-likes, such as
+ * `astropy.units.Quantity` and `np.ma.masked`.  These are used to us calling
+ * `__float__`/`__int__` for 0-D instances in many cases.
+ * Eventually, we may want to define this as wrong: They must use DTypes
+ * instead of (only) subclasses.  Until then, here as well as in
+ * `PyArray_AssignFromCache` (which already does this), we need to special
+ * case 0-D array-likes to behave like arbitrary (unknown!) Python objects.
+ *
+ * @param descr
+ * @param item
+ * @param value
+ * @return 0 on success -1 on failure.
+ */
+/*
+ * TODO: This function should possibly be public API.
+ */
+NPY_NO_EXPORT int
+PyArray_Pack(PyArray_Descr *descr, char *item, PyObject *value)
```

Comment:
Hmm, can you elaborate on this (or maybe I just need to step through the code)?  This should hit this code with the changes, if it does not, I missed a case where `dtype->setitem` should be replaced with `PyArray_Pack()`, since thats now the defined way to put a scalar into an array (I find it pretty fun that we never had a true correct way to do it).

I left out a few cases, on purpose, they seemed either corner case, or have well defined input, but this should not be one of them.

Reply:
After this code change :

```
>>> x = np.array(["qwdqwdqwd", "wqdqwd"])
>>> x[0] = np.datetime64("2020-01-01")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The string provided for NumPy ISO datetime formatting was too short, with length 9
```

on numpy 1.18.5

```
>>> import numpy as np
>>> x = np.array(["qwdqwdqwd", "wqdqwd"])
>>> x[0] = np.datetime64("2020-01-01")
>>> x
array(['2020-01-0', 'wqdqwd'], dtype='<U9')
```

Reply:
Aah, didnt really pay attention the error message. I guess this is considered one of the fixed bugs of this PR !?

Reply:
The reuslt is now consistent with `np.datetime64("2020-01-01").astype("U5"), whether the casting behaviour is ideal, is another question...

I guess this is reversed to the float64(NaN) case though, because its actually more strict now, hmmmm.  I could fix that, by defining that a datetime64 is a type known to string (and thus handled by `string->setitem`m which calls `str(...)` itself. Or just do that for all `NumPy scalars -> string` conversions, since at least the `__str__` dunder is pretty well defined...

Reply:
Oh, and I have to relax it for `np.array(np.datetime64("2020-01-01"), "U9")` as well...

Reply:
> The reuslt is now consistent with `np.datetime64("2020-01-01").astype("U5"), whether the casting behaviour is ideal, is another question...

aah got it.

> I could fix that, by defining that a datetime64 is a type known to string (and thus handled by string->setitemm which calls str(...) itself.

sorry not familiar with all the code here. how will this change look on PyArray_pack side ? currently it calls DATETIME_setitem in this line 

```
      if (tmp_descr->f->setitem(value, data, &arr_fields) < 0) {
```

will this change somehow ?

> Or just do that for all NumPy scalars -> string conversions, since at least the __str__ dunder is pretty well defined...

This sounds reasonable to me. So `np.datetime64("2020-01-01").astype("U5")` will also not error now right ? not sure about the backward compatibility and deprecation cycle though, since this is a behavior change.

Reply:
I just pushed that change, I am starting to be unwilling to squabble about whether or not this is small enough to just document as a change.  Normally, I would say deprecate, but I am starting to wear down, there are too many of these tiny things making it hard to focus on the big picture.  It is super difficult to fix the big picture while retaining behaviour that was never well defined :/.

(not to say it isn't important to undo it for now to be on the safe side)

Sorry: And as the explentation. For `PyArray_Pack()` the change in the code path is that it will hit `String->is_known_scalar_type(datetime64)` and unlike before it now returns `True` and thus it calls `String->setitem(datetime64)`, instead of using `Datetime->setitem(...).astype("S6")` internally.

Reply:
Thanks for the explanation, this sounds good to me,  and agreed we should get rid of the inconsistency with `np.datetime64("2020-01-01").astype("U5")`  in the future with astype sometime in the future.